### PR TITLE
Add plural fragment support to `observeFragment()`

### PR DESCRIPTION
**Pull request by XiNiHa**
This PR adds plural fragment support to `observeFragment()` (and `waitForFragmentData()` which uses `observeFragment()` under the hood)

It should be noted that the Relay environment currently doesn't support subscribing to plural fragment updates, so the implementation is done by subscribing to each selector and merging the states. Since the current implementation doesn't do any batching of notifications that are made from each selector subscription, it can be somewhat inefficient if the consumer doesn't batch on their own.

### CODE REVIEW

File: packages/relay-runtime/store/observeFragmentExperimental.js [hunk 1]
```diff
@@ -173,6 +170,67 @@ function observeSelector<TFragmentType: FragmentType, TData>(
...
+          if (state.state === 'ok') {
+            acc.value.push(state.value);
+            return acc;
+          } else {
+            return state;
+          }
+        },
+        {
+          state: 'ok',
+          // $FlowFixMe[incompatible-cast]: TData is enforced to be an array
```
Comment (XiNiHa) [inline packages/relay-runtime/store/observeFragmentExperimental.js:211]: Not sure why this should be marked as `$FlowFixMe` when `TData` has `: Array<mixed>` enforcement

**Comment by captbaritone** [comment]
Exciting! I'll take a look tomorrow.

File: packages/relay-runtime/store/observeFragmentExperimental.js [hunk 2]
```diff
@@ -173,6 +170,67 @@ function observeSelector<TFragmentType: FragmentType, TData>(
...
+  const snapshots = fragmentSelector.selectors.map(selector =>
+    environment.lookup(selector),
+  );
+
+  return Observable.create(sink => {
+    const states = snapshots.map((snapshot, index) =>
+      snapshotToFragmentState(
+        environment,
+        fragmentNode,
+        fragmentSelector.selectors[0].owner,
```
Comment (captbaritone) [inline packages/relay-runtime/store/observeFragmentExperimental.js:190]: Either use the `index` here or omit it from the callback arguments on line 186

File: packages/relay-runtime/store/observeFragmentExperimental.js [hunk 3]
```diff
@@ -173,6 +170,67 @@ function observeSelector<TFragmentType: FragmentType, TData>(
...
+        snapshot,
+      ),
+    );
+
+    const reduceState = () =>
+      states.reduce<FragmentState<TData>>(
+        (acc, state) => {
+          if (acc.state !== 'ok') {
+            return acc;
```
Comment (captbaritone) [inline packages/relay-runtime/store/observeFragmentExperimental.js:199]: I wonder if this would be easier to read as a for loop with a `break`? It's a little hard to follow right now, and I think right now we still have to continue iterating the states even after we hit one that's not "ok". Would be nicer if we could just bail and return the loading or error state right away.

Maybe we could define a generic function (outside of this closure) which does this transformation:

```ts
function namingIsHard<T>(states: Array<FragmentState<T>>): FragmentState<Array<T>> {
  const values: Array<T> = [];
  for(let i = 0; i < states.length; i++) {
    const state = states[i];
    if(state.state === 'ok') {
      values.push(state.value);
    } else {
      return state;
    }
  }
  return { state: "ok", value: values };
}
```

File: packages/relay-runtime/store/observeFragmentExperimental.js [hunk 4]
```diff
@@ -173,6 +170,67 @@ function observeSelector<TFragmentType: FragmentType, TData>(
...
+  environment: IEnvironment,
+  fragmentNode: Fragment<TFragmentType, TData>,
+  fragmentSelector: PluralReaderSelector,
+): Observable<FragmentState<TData>> {
+  const snapshots = fragmentSelector.selectors.map(selector =>
+    environment.lookup(selector),
+  );
+
+  return Observable.create(sink => {
+    const states = snapshots.map((snapshot, index) =>
```
Comment (captbaritone) [inline packages/relay-runtime/store/observeFragmentExperimental.js:186]: Might be worth a comment up here explaining that `states` is mutable and its values will get updated in place overtime

File: packages/relay-runtime/store/observeFragmentExperimental.js [hunk 5]
```diff
@@ -173,6 +170,67 @@ function observeSelector<TFragmentType: FragmentType, TData>(
...
+
+    const subscriptions = snapshots.map((snapshot, index) =>
+      environment.subscribe(snapshot, latestSnapshot => {
+        states[index] = snapshotToFragmentState(
+          environment,
+          fragmentNode,
+          fragmentSelector.selectors[0].owner,
+          latestSnapshot,
+        );
+        sink.next(reduceState());
```
Comment (captbaritone) [inline packages/relay-runtime/store/observeFragmentExperimental.js:226]: Maybe a comment here calling out that this does not batch updates, and if a store update impacting multiple items in the list is published, we will notify the subscriber multiple times?

File: packages/relay-runtime/store/__tests__/observeFragment-test.js [hunk 6]
```diff
@@ -337,6 +337,189 @@ test('read deferred fragment', async () => {
...
+      {
+        error: new Error(
+          "Relay: Resolver error at path 'always_throws' in 'observeFragmentTestResolverErrorWithPluralThrowOnFieldErrorFragment'.",
+        ),
+        state: 'error',
+      },
+    ]);
+  });
+});
+
```
Comment (captbaritone) [inline packages/relay-runtime/store/__tests__/observeFragment-test.js:522]: Can we add some additional tests:

1. A test demonstrating a store update/publish that touches multiple items in the list triggering multiple observable updates?
2. A test or two demonstrating that `waitForFragmentData` also works with plural fragments?

**Review by captbaritone — COMMENTED** [review]
Overall this looks great! Just a few nits.

Could you also add some notes to the docs about the behavior for plural fragments and letting folks know they can debounce if they want to batch?

File: packages/relay-runtime/store/observeFragmentExperimental.js (hunk 3 above)
Comment (XiNiHa) [reply packages/relay-runtime/store/observeFragmentExperimental.js]: https://github.com/facebook/relay/pull/4862/commits/4a5629881cafb7d8b53fc9d18314b9afe875f749

File: packages/relay-runtime/store/observeFragmentExperimental.js (hunk 2 above)
Comment (XiNiHa) [reply packages/relay-runtime/store/observeFragmentExperimental.js]: https://github.com/facebook/relay/pull/4862/commits/70f047e6725e7aa9e60277e6b15054b7405e07ee

File: packages/relay-runtime/store/observeFragmentExperimental.js (hunk 4 above)
Comment (XiNiHa) [reply packages/relay-runtime/store/observeFragmentExperimental.js:187]: https://github.com/facebook/relay/pull/4862/commits/06903bfa7a0030c3d5569051b47479e968627bf8

File: packages/relay-runtime/store/observeFragmentExperimental.js (hunk 5 above)
Comment (XiNiHa) [reply packages/relay-runtime/store/observeFragmentExperimental.js]: https://github.com/facebook/relay/pull/4862/commits/06903bfa7a0030c3d5569051b47479e968627bf8

File: packages/relay-runtime/store/__tests__/observeFragment-test.js (hunk 6 above)
Comment (XiNiHa) [reply packages/relay-runtime/store/__tests__/observeFragment-test.js:522]: https://github.com/facebook/relay/pull/4862/commits/e596faac720d127baea60f84ee2c9b286ffbc72d

**Review by captbaritone — APPROVED** [review]
This looks so good. Thanks for filling this gap and making this API more usable for everyone!
